### ISSUE TITLE
from_pandas fails to clone data

### ISSUE THREAD

**Issue by desprado2** [pola-rs/polars#12582]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import numpy as np
import pandas as pd
import polars as pl

assert pl.__version__ == "0.19.14"

id = np.range(6 * 10**6)
values = np.random.randn(6 * 10**6)

pd_df = pd.DataFrame({"val": values}, index=pd.Index(id, name="id"))
pl_df = pl.DataFrame(pd_df, include_index=True).clone() # even if I clone the dataframe, the problem still happens.

pd_df["val"] += 3 # the polars dataframe should not change at all

assert not pd_df["val"].to_list() == pl_df["val"].to_list() # assertion failed
```

### Issue description

As in API reference, `from_pandas` should clone data, as well as `clone`. However, after I modify `pd_df`, the generated polars dataframe `pl_df` is also modified.

### Expected behavior

The modification of the source dataframe should not affect new dataframe.

### Installed versions

<details>

```
polars = 0.19.14
pandas = 2.0.2
numpy = 1.24.4
```

</details>

**Comment by ritchie46** [comment]
I have to think about this one a bit.

Pandas is happily giving out numpy arrays and overwriting data in that numpy array later on.

```python
id = np.arange(6)
values = np.random.randn(6)

pd_df = pd.DataFrame({"val": values}, index=pd.Index(id, name="id"))

a = pd_df["val"].to_numpy()
print(a)

pd_df["val"] += 3 # changes the numpy array

print(a)
```

```
[-0.32757808  0.4235578  -0.82854383  1.37921495  0.06696384 -0.18274764]
[2.67242192 3.4235578  2.17145617 4.37921495 3.06696384 2.81725236]
```

**Comment by MarcoGorelli** [comment]
copy-on-write isn't the default in pandas (yet - will be in 3.0), but if you enable it you'll get:
```python
In [1]: with pd.option_context("mode.copy_on_write", True):
   ...:     id = np.arange(6)
   ...:     values = np.random.randn(6)
   ...:     pd_df = pd.DataFrame({"val": values}, index=pd.Index(id, name="id"))
   ...:     a = pd_df["val"].to_numpy()
   ...:     print(a)
   ...:     pd_df["val"] += 3 # doesn't change the numpy array
   ...:     print(a)
   ...:
[-0.4739135   0.35082205 -0.78352693  1.15089348  0.57343854 -1.25351246]
[-0.4739135   0.35082205 -0.78352693  1.15089348  0.57343854 -1.25351246]
```

**Comment by MarcoGorelli** [comment]
For now (until pandas 3.0) I'd suggest doing
```python
pl.from_pandas(pd_df.copy(deep=True), include_index=True)
```

**Comment by desprado2** [comment]
> For now (until pandas 3.0) I'd suggest doing
> 
> ```python
> pl.from_pandas(pd_df.copy(deep=True), include_index=True)
> ```

Thank you for pointing out. Shall we add this suggestion to doc?

**Comment by tunkill** [comment]
Even with Polars 1.30.0, the most recent version, I'm still encountering this problem, especially with numeric column values. It seems the Polars DataFrame is unexpectedly affected after modifying the original Pandas DataFrame. I'm wondering if this is an issue that should be resolved internally within Polars – perhaps by having the pl.from_pandas() function internally implement pd_df.copy(deep=True).

**Comment by MarcoGorelli** [comment]
if it did that, then someone else would file as issue saying "`from_pandas` copies data unnecessarily"

i'd suggest either enabling copy-on-write in pandas, or using `.copy` on the pandas dataframe

**Comment by tunkill** [comment]
> if it did that, then someone else would file as issue saying "`from_pandas` copies data unnecessarily"
> 
> i'd suggest either enabling copy-on-write in pandas, or using `.copy` on the pandas dataframe

If that's the case, I was wondering if it might be helpful to clarify the documentation for pl.from_pandas(). The current docstring mentions: 'Construct a Polars DataFrame or Series from a pandas DataFrame, Series, or Index. This operation clones data.' From my experience, especially with numeric column values, this can be a bit confusing, as the data doesn't always seem to be cloned; it appears to be shared instead. Perhaps, as an alternative, a parameter could be considered for pl.from_pandas() to allow users to explicitly control this behavior? Thank you very much for your consideration.

**Comment by MarcoGorelli** [comment]
Ah thanks - yeah the docstring should probably be updated to say "This operation may clone data", and advise what to do if you want to ensure that data is cloned

**Comment by cmdlineluser** [comment]
I did notice a somewhat related PR:
 
- https://github.com/pola-rs/polars/pull/22942

Testing `pl.DataFrame(pd_df.reset_index())` - it does not seem to be affected by this issue?

**Comment by MarcoGorelli** [comment]
yeah `reset_index` makes a copy. it can be real performance footgun if you're not careful (side note, i think a big part of the perf improvements on the pandas side reported in https://plotly.com/blog/chart-smarter-not-harder-universal-dataframe-support/ were due to removing unnecessary `reset_index`s..)
